### PR TITLE
test(install-ps1): surface powershell timeouts and add single retry on delegation

### LINKED ISSUE #1445: test(install-ps1): the PowerShell dry-run delegation test hit the 90 s budget on the windows-latest runner for main at 3b5c0f36, green on rerun
## What happened

On the CI run for main at 3b5c0f36 (the squash of #1439), the job `Test (windows-latest, Node 22.x, npm)` failed in `tests/scripts/install-ps1.test.js`:

```
✗ delegates to the Node installer and preserves dry-run output
    Error: Expected values to be strictly equal:
    1 !== 0
```

The same job, on the same tree (the PR head 5da6e6e3 sits directly on top of main, so the squash has an identical tree), had passed minutes earlier, with the whole `install.ps1` suite taking about three seconds (17:41:43 to 17:41:46 UTC). On main the suite ran from roughly 17:52:56 to 17:57:08 UTC before failing. A rerun of the failed job (attempt 2 of run 34772670974) passed.

## Reading

`1 !== 0` with an empty stderr is the mute signature `tests/fixtures/subprocess-timeouts.js` already documents: `execFileSync` killed the PowerShell dry run at `FULL_INSTALL_TIMEOUT_MS` (90 s), its null exit status became `code: 1` in the test helper, and the assertion message got an empty stderr. So the dry run, which normally completes in about a second on that runner, did not return within 90 s. Nothing in the diff of #1439 adds work to a dry run against an empty temp tree (the shape scan only `lstat`s planned paths that do not exist yet), and the rerun on the same commit confirms the code is not the cause. This looks like a cold or stalled PowerShell start on the hosted Windows runner.

## What would make the next occurrence cheaper to read

- The helper in `tests/scripts/install-ps1.test.js` could surface the timeout distinctly: when `error.code === 'ETIMEDOUT'` or `error.signal` is set, fail with a message that names the budget and the elapsed time instead of the bare `1 !== 0`.
- Optionally, retry the delegation once when the first attempt hits the budget, so a single cold start does not turn a green tree red.

Tracking here so the pattern is visible if it repeats. Run: https://github.com/Fmarzochi/EGC/actions/runs/34772670974

### PR DESCRIPTION
**Pull request by Akisolu**
Draft PR for early feedback / clarification on the `install-ps1` timeout handling.

## What Changed
- Updated `run()` helper in `tests/scripts/install-ps1.test.js` to measure execution time (`elapsedMs`) and preserve `code: null` when a subprocess is killed via timeout or signal (`timedOut: true`).
- Updated the PowerShell dry-run delegation test to perform a single retry if the initial attempt hits a timeout (e.g., hosted runner cold start) before failing.
- Improved assertion failure output on timeout to explicitly report elapsed time versus budget.

## Why This Change
Prevents intermittent CI failures on `windows-latest` caused by PowerShell cold starts (as tracked in #1445) while ensuring the test failure message remains informative instead of a mute `1 !== 0`.
Close #1445 

## Testing Done
- [x] Manual testing completed
- [ ] Automated tests pass locally (`node tests/run-all.js`)
- [x] Edge cases considered and tested
- [ ] If this PR touches a file shared across concurrent EGC processes (encryption key, state files, install-state, lockfiles under `~/.egc/`), a concurrent-access test was added or updated

## Type of Change
- [ ] `fix:` Bug fix
- [ ] `feat:` New feature
- [ ] `refactor:` Code refactoring
- [ ] `docs:` Documentation
- [x] `test:` Tests
- [ ] `chore:` Maintenance/tooling
- [ ] `ci:` CI/CD changes

## Security & Quality Checklist
- [x] JSON files validate cleanly
- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] Shell scripts pass shellcheck (if applicable)
- [x] Pre-commit hooks pass locally (if configured)
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## Documentation
- [ ] Updated relevant documentation
- [x] Added comments for complex logic
- [ ] README updated (if needed)

---

### Question for @Fmarzochi

Hey Felipe! Opening this as a **Draft PR** to double-check a detail with you:

In the `run()`  helper 0bd477603372cd115965db18d853b3d48c64837f, when `execFileSync` throws on a timeout, `error.status` is `null`. I've set `code: isTimeout ? null : (error.status ?? 1)` so that `timedOut: true` properly retains `code: null`. 

When executing the retry 5e55ee8e3107684555f5d39f4531cc7f071c6a04, I repeated the code from lines 282–285.

Could you take a quick look at the inline diff for `tests/scripts/install-ps1.test.js` and confirm if this shape for `run()` and the single-retry structure matches what you had in mind before I mark this as ready for review?

Besides mentioning that I don't consider my current work finished, I haven't yet run `node tests/scripts/install-ps1.test.js` to test the implementation.

---
## Summary by cubic
Fixes `install-ps1` test timeout handling so Windows runner cold starts are retried once and timeout failures are easier to diagnose. `run()` now returns `code: null` with `timedOut: true` when the subprocess times out or is killed by a signal, and reports `elapsedMs`. The dry-run delegation test retries once after a timeout, then asserts on the retried result, and a new test covers `describeFailure` classification of timeouts, signals, and exit codes.

<sup>Written for commit d870c77f728a02a4870c43cf414288ffb9434528. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/Fmarzochi/EGC/pull/1447?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Tests**
  - Improved reliability of PowerShell installation checks by classifying timeouts, terminated processes, and exit-code failures consistently.
  - Added elapsed-time details and clearer failure reporting to support troubleshooting.
  - Added coverage for failure classification.
  - Added a single retry for delegation checks when the initial dry run times out or is signal-terminated, reducing intermittent test failures.

### CODE REVIEW

**Comment by Fmarzochi** [comment]
Thanks for the fast draft, Akisolu. The shape of `run()` is the one I meant: `code: isTimeout ? null : (error.status ?? 1)` plus `timedOut` and `elapsedMs` is exactly right. Two runtime defects and one gap stand between this and the merge; I ran the patterns to be sure, and the repo's own eslint reports the first two (`npx eslint tests/scripts/install-ps1.test.js`, which is what the red Lint job runs).

1. **`startTime` is declared inside the `try`, so it does not exist in the `catch`.** Every failing run, timeout or not, now throws `ReferenceError: startTime is not defined` instead of returning a result (eslint: `no-undef` at line 67). Move `const startTime = Date.now();` above the `try`.

2. **The retry is thrown away.** Inside `if (result.timedOut) { ... }` you declare a new `const result`, which shadows the outer one; the second run executes and the assertions below still read the first, timed-out attempt (eslint: `no-unused-vars` at line 288). Write it as one variable and one call site:

   ```js
   const delegate = () => run(powerShellCommand, ['--target', 'cursor', '--dry-run', 'typescript'], { cwd: projectDir, homeDir });
   let result = delegate();
   if (result.timedOut) {
     console.log(`  - retrying the delegation: the first attempt hit the ${FULL_INSTALL_TIMEOUT_MS} ms budget after ${result.elapsedMs} ms`);
     result = delegate();
   }
   ```

   That also removes the duplicated argument list you mentioned.

3. **The failure is still mute.** With `code: null` and an empty stderr, `assert.strictEqual(result.code, 0, result.stderr)` prints `null !== 0`, which is the old symptom with a different number. Assert the timeout first, with the numbers in the message:

   ```js
   assert.ok(!result.timedOut, `the PowerShell dry run exceeded FULL_INSTALL_TIMEOUT_MS on the retry as well: ${result.elapsedMs} ms of ${FULL_INSTALL_TIMEOUT_MS} ms`);
   ```

   and keep the existing assertions after it.

4. **DCO.** Neither commit carries `Signed-off-by`, so the DCO job is red. Run `git config core.hooksPath .githooks` once in your clone so every new commit is signed automatically, then re-sign the two existing commits (`git rebase --signoff main` does both at once) and update your branch. While there, `test(install-ps1):` as the commit type matches the PR title better than `fix(tests):`.

5. Small style points the linter will not catch: `elapsedMs: elapsedMs` can be the shorthand `elapsedMs`, and `if (result.timedOut){` wants the space before the brace.

6. Proof: a Linux or macOS box without `pwsh` skips both PowerShell tests (mine does, 13 pass and 2 skipped), so the Windows lanes in CI and your own Windows machine are the real proof. Please run `node tests/scripts/install-ps1.test.js` there, once with a tiny budget to see the new message, then with the budget back at 90 s, and tick the checklist when it passes. SonarCloud already reports zero new issues on this head, so once eslint and DCO are green the bar is met.

Ping me when it is out of draft and I will run the same checks again the same day.

**Comment by Akisolu** [comment]
Hi @Fmarzochi, I've updated the PR with all the requested fixes:

1. **Helper `run()` updated:** `const startTime = Date.now();` moved outside the `try` block, and `elapsedMs` shorthand applied.
2. **Clean retry logic:** Implemented `delegate()` closure to avoid variable shadowing, added the retry log, and inserted the descriptive `assert.ok(!result.timedOut, ...)` check before standard code assertions.
3. **Linting verified:** `npx eslint tests/scripts/install-ps1.test.js` now runs clean with zero warnings/errors.
image

4. **DCO & Hooks:** Configured `.githooks` and rebased with `--signoff`.
5. **Local Windows verification:** Tested with a forced short budget to confirm the retry logging and timeout error formatting, then verified the entire suite passes with the restored 90 s budget.
image
image

**Note:**
When testing locally on Windows with a temporary 50 ms budget:
- The delegation test logged the retry and surfaced the new timeout message as intended.
- The `exposes the corrected Gemini target help text` test also timed out, failing with the standard assertion since retry logic is scoped strictly to delegation.

With the 90 s budget restored, the full suite passes cleanly in ~3 s. Leaving that observation here in case you'd like to address that second test as well, otherwise it's ready for final review!

The PR is out of draft and ready for your final review!

**Comment by Fmarzochi** [comment]
Round two verified on my side: eslint clean, the thirteen non-PowerShell tests pass on Linux with the two PowerShell ones skipping as expected, SonarCloud reports zero new issues on 2b8c4389, all four commits carry the sign-off, and your Windows run with the forced budget shows the retry log and the new message doing their job. Good round.

One reviewer finding stands, and it is right, so one more small push before the merge:

1. **A signal is not a timeout.** `Boolean(error.signal)` makes any child killed by a signal (an OOM kill, a cancelled runner) look like it exceeded the budget, and the message would then claim `FULL_INSTALL_TIMEOUT_MS` was exceeded when it was not. Classify precisely and keep the retry for both, since both are environmental deaths:

   ```js
   const timedOut = error.code === 'ETIMEDOUT';
   const signal = error.signal ?? null;
   return { code: error.status ?? null, stdout: error.stdout || '', stderr: error.stderr || '', timedOut, signal, elapsedMs };
   ```

   In the delegation test retry when `result.timedOut || result.signal`, and let the log and the final assertion name what happened: the budget and the elapsed time when `timedOut`, the signal and the elapsed time when `signal`.

2. **Make the classification testable where every runner can run it.** Pull the `catch` body into a small pure function, for example `describeFailure(error, elapsedMs)`, and add one test in this file that feeds it three fake errors: `{ code: 'ETIMEDOUT', signal: 'SIGTERM', status: null }` gives `timedOut: true`; `{ signal: 'SIGKILL', status: null }` gives `timedOut: false, signal: 'SIGKILL', code: null`; `{ status: 2 }` gives `code: 2`. That test does not need PowerShell, so Linux and macOS lanes cover the logic too, and it answers the "regression test for signal termination" the reviewer asked for without spawning anything.

On the help-text test: leave it. Its budget is the 30 s CLI one, a 50 ms budget will time out anything, and the cold start we saw in CI hit the dry run, not `--help`. If the pattern ever shows up on that test we open a follow-up with the log, the same way #1445 was opened.

Same proof as before on your side (eslint, the suite on Windows with the forced budget once, then the 90 s budget back) and I run mine the same day.

File: tests/scripts/install-ps1.test.js [hunk 1]
```diff
@@ -63,10 +64,14 @@ function run(powerShellCommand, args = [], options = {}) {
 
     return { code: 0, stdout, stderr: '' };
   } catch (error) {
+    const elapsedMs = Date.now() - startTime;
+    const isTimeout = error.code === 'ETIMEDOUT' || Boolean(error.signal);
```
[the remark this answers is not in the document]
Comment (Akisolu) [reply tests/scripts/install-ps1.test.js]: Refactored error classification to keep diagnostics accurate d870c77:

- Extracted failure parsing into a pure `describeFailure(error, elapsedMs)` helper.
- Strictly separated `timedOut` (`error.code === 'ETIMEDOUT'`) from OS signals (`error.signal ?? null`).
- Retried delegation on either environmental death (`timedOut` or `signal`), but added distinct assertions to properly attribute failures to either budget timeouts or process signals (e.g., `SIGKILL`).
- Added a standalone unit test validating `describeFailur

[the remark this answers is not in the document]
Comment (Fmarzochi) [reply tests/scripts/install-ps1.test.js]: Addressed in 7986a227 and d870c77f: describeFailure sets timedOut only on ETIMEDOUT, returns the signal separately with code null, the delegation retries on either and names the cause in the log and in the assertion, and a three-case unit test covers the classification on every runner. Resolving.

[the remark this answers is not in the document]
Comment (Fmarzochi) [reply tests/scripts/install-ps1.test.js]: Addressed in 7986a227 and d870c77f: describeFailure sets timedOut only on ETIMEDOUT, returns the signal separately with code null, the delegation retries on either and names the cause in the log and in the assertion, and a three-case unit test covers the classification on every runner. Resolving.

**Comment by Akisolu** [comment]
Hi Felipe.

All items from round two are complete:

1. **Precise Failure Classification:** Refactored error handling into `describeFailure(error, elapsedMs)`. Strictly separated `timedOut` (`error.code === 'ETIMEDOUT'`) from `signal` (`error.signal ?? null`), keeping retry logic for both environmental deaths with explicit, distinct assertion messages.
2. **Pure Unit Test:** Extracted `describeFailure` and added a standalone test covering `ETIMEDOUT`, `SIGKILL`, and non-zero exit codes. It executes cleanly across all OS runners without spawning subprocesses.
3. **Verification:** `npx eslint te
Pushed, signed via DCO, and ready for the your review!sts/scripts/install-ps1.test.js` passes clean, the test suite passes on Windows.

File: tests/scripts/install-ps1.test.js [hunk 2]
```diff
@@ -41,13 +41,27 @@ function resolvePowerShellCommand() {
   return null;
 }
 
+function describeFailure(error, elapsedMs) {
+  const timedOut = error.code === 'ETIMEDOUT';
+  const signal = error.signal ?? null;
+  return { 
+    code: error.status ?? null,
```
[the remark this answers is not in the document]
Comment (Fmarzochi) [reply tests/scripts/install-ps1.test.js:48]: Settled, no change needed. The discriminators are timedOut and signal, not code: a spawn failure that never produced an exit code is honestly reported as code null, and inventing a 1 there would fake an exit code that did not happen. The assertion still prints the stderr of such a failure (an ENOENT names the missing binary), so nothing is mute. This is the shape I asked for in review.

**Comment by Fmarzochi** [comment]
Round three verified on my side: eslint clean, fourteen tests pass on Linux with the new classification test among them and the two PowerShell tests skipping as expected, SonarCloud reports zero new issues on d870c77f, all six commits carry the sign-off, and both reviewer threads are resolved by the code. That is the whole bar.

Merging with your authorship. Thank you, Akisolu: fourth PR, three rounds in one evening, and a test file that now explains itself when a runner misbehaves.

**Comment by Akisolu** [comment]
Thanks Felipe! Glad we got this properly classified and testable across all runners. 

It was a great solid round of refactoring. See you in the next one!

**Comment by Fmarzochi** [comment]
@Akisolu  This round went excellently.
I’m grateful for your help and we’re moving toward the release soon...
